File: 정종길/프로그래머스/카카오 1회독/2022 KAKAO BLIND RECRUITMENT/파괴되지 않은 건물.java
import java.util.*;

class Solution {
    public int solution(int[][] board, int[][] skill) {
        int answer = 0;
        int r = board.length;
        int c = board[0].length;
        int[][] prefix = new int[r + 1][c + 1];

        for(int [] order : skill){
            if(order[0] == 1){
                prefix[order[1]][order[2]] += -order[5];
                prefix[order[1]][order[4] + 1] += order[5];
                prefix[order[3] + 1][order[2]] += order[5];
                prefix[order[3] + 1][order[4] + 1] += -order[5];
            } else {
                prefix[order[1]][order[2]] += order[5];
                prefix[order[1]][order[4] + 1] += -order[5];
                prefix[order[3] + 1][order[2]] += -order[5];
                prefix[order[3] + 1][order[4] + 1] += order[5];
            }
        }


        for(int i = 0 ; i < r; i ++){
            for(int j = 0 ; j < c - 1; j ++){
                prefix[i][j + 1] += prefix[i][j];
            }
        }

        for(int j = 0 ; j < c; j ++){
            for(int i = 0 ; i < r - 1; i ++){
                prefix[i + 1][j] += prefix[i][j];
            }
        }

        for(int i = 0 ; i < r; i ++){
            for(int j = 0 ; j < c; j ++){
                if(board[i][j] + prefix[i][j] > 0){
                    answer += 1;
                }
            }
        }

        return answer;
    }
}

File: 정종길/프로그래머스/카카오 2회독/2022 KAKAO BLIND RECRUITMENT/파괴되지 않은 건물.py
def solution(board, skill):
    answer = 0
    prefix_sums = [[0] * (len(board[0]) + 1) for _ in range(len(board) + 1)]
    for t, r1, c1, r2, c2, d in skill:
        if t == 1:
            prefix_sums[r1][c1] += -d
            prefix_sums[r2 + 1][c1] += d
            prefix_sums[r1][c2 + 1] += d
            prefix_sums[r2 + 1][c2 + 1] += -d
        else:
            prefix_sums[r1][c1] += d
            prefix_sums[r2 + 1][c1] += -d
            prefix_sums[r1][c2 + 1] += -d
            prefix_sums[r2 + 1][c2 + 1] += d

    for i in range(len(prefix_sums) - 1):
        for j in range(len(prefix_sums[0]) - 2):
            prefix_sums[i][j + 1] += prefix_sums[i][j]
    
    for j in range(len(prefix_sums[0]) - 1):
        for i in range(len(prefix_sums) - 2):
            prefix_sums[i + 1][j] += prefix_sums[i][j]
    
    for i in range(len(board)):
        for j in range(len(board[0])):
            if board[i][j] + prefix_sums[i][j] > 0:
                answer += 1
    return answer